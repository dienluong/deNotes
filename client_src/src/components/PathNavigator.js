import React from 'react';

class PathNavigator extends React.Component {
  render() {
    const path = this.props.path.length ?
      this.props.path.map((step, idx) => <span key={ idx }>\{ step }</span>) :
      <span>\</span>;
    return (
      <div>
        { path }
      </div>
    );
  }
}

export default PathNavigator;
