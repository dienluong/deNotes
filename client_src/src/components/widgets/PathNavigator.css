.lined {
  display: inline-block;
}

.active-segment {
  color: pink;
  font-weight: bold;
}
