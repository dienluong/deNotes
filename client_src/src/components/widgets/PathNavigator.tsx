import React from 'react';
import PathSegment from './PathSegment';
import styles from './PathNavigator.module.css';

// Types
type PropsT = {
  path: string[];
  activeSegmentIdx: number;
  onClick: (params: { idx: number }) => any;
}

function PathNavigator({ path, activeSegmentIdx, onClick: clickHandler }: PropsT) {
  const pathSegments = path.length ?
    path.map((step, idx) =>
      <PathSegment
        className={ styles.lined + ' ' + (idx === activeSegmentIdx ? styles['dnt__pathnav-segment--active'] : styles['dnt__pathnav-segment']) }
        key={ idx }
        label={ step }
        onClick={ () => clickHandler({ idx }) }
      />) :
    [];

  return (
    <nav className={ styles.dnt__pathnav }>
      <ul>
        { pathSegments }
      </ul>
    </nav>
  );
}

export default PathNavigator;
