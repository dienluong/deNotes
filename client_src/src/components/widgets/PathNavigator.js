import React from 'react';
import PathSegment from './PathSegment';
import './PathNavigator.css';

function PathNavigator({ path, onClick: clickHandler }) {
  const pathSegments = path.length ?
    path.map((step, idx) =>
      <PathSegment
        className='lined'
        key={ idx }
        label={ step }
        onClick={ () => clickHandler({ idx }) }
      />) :
    [];

  return (
    <nav>
      <ul>
        { pathSegments }
      </ul>
    </nav>
  );
}

export default PathNavigator;
