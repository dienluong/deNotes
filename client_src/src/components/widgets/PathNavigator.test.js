import PathNavigator from './PathNavigator';
import React from 'react';
import { render, cleanup, fireEvent } from 'react-testing-library';
import 'jest-dom/extend-expect';
const ACTIVE_SEGMENT_CLASS = 'active-segment';

afterEach(cleanup);

it('renders an element composed of segments w/ labels and with specified segment as "active"', () => {
  const props = {
    path: ['label1', 'label2', 'label3'],
    onClick: jest.fn(),
    activeSegmentIdx: 1,
  };

  const { queryAllByText, container } = render(<PathNavigator { ...props } />);

  // expect rendered component to have all specified labels
  props.path.forEach(label => expect(queryAllByText(label, { exact: false })).toHaveLength(1));

  // expect one segment as active
  const activeSegment = container.getElementsByClassName(ACTIVE_SEGMENT_CLASS);
  expect(activeSegment).toHaveLength(1);
  expect(activeSegment[0]).toHaveTextContent(props.path[props.activeSegmentIdx]);
});

it('each segment of rendered component reacts to click event', () => {
  const props = {
    path: ['label1', 'label2', 'label3'],
    onClick: jest.fn(),
    activeSegmentIdx: 1,
  };

  const { queryAllByText } = render(<PathNavigator { ...props } />);
  const allSegments = queryAllByText(/label/);

  // click on each of the segments
  allSegments.forEach((seg, idx) => {
    fireEvent.click(seg);
    expect(props.onClick).lastCalledWith({ idx });
  });

  expect(props.onClick).toBeCalledTimes(props.path.length);
});
