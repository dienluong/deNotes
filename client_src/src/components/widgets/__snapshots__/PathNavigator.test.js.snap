// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an element composed of segments w/ labels and with specified segment as "active" 1`] = `
<div>
  <nav
    class="dnt__pathnav"
  >
    <ul>
      <li
        class="lined"
      >
        <span>
          label1
          &gt;
        </span>
      </li>
      <li
        class="lined dnt__pathnav-segment--active"
      >
        <span>
          label2
          &gt;
        </span>
      </li>
      <li
        class="lined"
      >
        <span>
          label3
          &gt;
        </span>
      </li>
    </ul>
  </nav>
</div>
`;
